### PR TITLE
Multivariate Gaussian Proposal

### PR DESCRIPTION
**Pull request by icantsayrural**
For the first pass, assumes diagonal covariance matrix.

There is also a small rounding error from calculating the log probability as a sum of Univariate Gaussians rather than one Multivariate (i.e. `log(.).sum() + log(.).sum()` instead of `(log(.) + log(.)).sum()`).

Next steps:
- Generalise to non-diagonal covariance matrix
- Adaptive step sizes
- More general, non-Gaussian multivariate proposal (i.e. a mixture of different proposals)
- More sensible default than `prior`

### CODE REVIEW

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/MultivariateGaussianProposalDistributionTest.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    @Before
+    public void setRandomSeed() {
+        KeanuRandom.setDefaultRandomSeed(0);
+    }
+
+    @Before
+    public void setUpProposal() {
+        when(vertex1.getValue()).thenReturn(currentStateForVertex1);
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/MultivariateGaussianProposalDistributionTest.java]: Lovely mocking but do you need it? What's wrong with just creating an actual GaussianVertex?

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/MultivariateGaussianProposalDistributionTest.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        proposal = new Proposal();
+        proposal.setProposal(vertex1, proposedStateForVertex1);
+        proposal.setProposal(vertex2, proposedStateForVertex2);
+    }
+
+    @Test
+    public void theLogProbAtToIsMultivariateGaussian() {
+        double logProb = proposalDistribution.logProbAtToGivenFrom(proposal);
+        DoubleTensor mu = DoubleTensor.concat(currentStateForVertex1, currentStateForVertex2);
+        DoubleTensor cov = DoubleTensor.concat(sigmaForVertex1, sigmaForVertex2).diag().pow(2.);
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/MultivariateGaussianProposalDistributionTest.java]: Not that it matters much here but `pow(2).diag()` is preferable?

File: keanu-python/tests/test_sampling.py [hunk 3]
```diff
@@ -176,6 +176,14 @@ def test_can_specify_a_gaussian_proposal_distribution(net: BayesNet) -> None:
     generate_samples(net=net, sample_from=net.iter_latent_vertices(), sampling_algorithm=algo)
 
 
+def test_can_specify_a_multivariate_gaussian_proposal_distribution(net: BayesNet) -> None:
+    algo = MetropolisHastingsSampler(
```
Comment (gordoncaleb) [inline keanu-python/tests/test_sampling.py]: Can you add a test to `MetropolisHastingsTest.java` that uses the proposal distribution? It should be good for the Gaussian sum test cast.

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/MultivariateGaussianProposalDistributionTest.java (hunk 1 above)
Comment (icantsayrural) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/MultivariateGaussianProposalDistributionTest.java]: Copied it from `GaussianProposalDistribution`

Comment (icantsayrural) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/proposal/MultivariateGaussianProposalDistributionTest.java]: I assumed the mock was to make it distribution agnostic?

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/proposal/MultivariateGaussianProposalDistribution.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        return proposal;
+    }
+
+    @Override
+    public <T> double logProb(Probabilistic<T> variable, T ofValue, T givenValue) {
+        if (!(ofValue instanceof DoubleTensor)) {
+            throw new ClassCastException("Only DoubleTensor values are supported - not " + ofValue.getClass().getSimpleName());
+        }
+
+        ContinuousDistribution proposalDistribution = Gaussian.withParameters((DoubleTensor) ofValue, sigmas.get(variable));
```
Comment (christophernorth) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/proposal/MultivariateGaussianProposalDistribution.java]: nit: Given we know this is a Gaussian - do we need to refer to it via the ContinuousDistribution interface?

File: release_notes.md [hunk 5]
```diff
@@ -2,6 +2,9 @@
 
 ### Common
 
+* Added `MultivariateGaussianProposalDistribution`. It assumes a diagonal covariance matrix.
+  * Example in Python:
+    * `MetropolisHastingsSampler(proposal_distribution='multivariate_gaussian', latents=[v1, v2], proposal_distribution_sigma=[1., np.array([[1., 2.], [3., 4.]]))`
```
Comment (christophernorth) [inline release_notes.md]: Do we need to add any of this to the shinydocs?

**Comment by icantsayrural** [comment]
> have we spoken to the requester to ensure that a diagonal covariance matrix is sufficient?

Yeah, I've chatted with Rob about it. He said it'll still be useful to be able to specify a different sigma for each vertex.

File: release_notes.md (hunk 5 above)
Comment (icantsayrural) [reply release_notes.md]: It'll be in Python docs

**Comment by icantsayrural** [comment]
Merged `MultivariateGaussianProposal` and `GaussianProposal` because if we assume a diagonal covariance matrix, it's just a generalised Gaussian proposal.

**Pull request merged**